### PR TITLE
Fix kafka version comparing when message format is used instead

### PR DESCRIPTION
**Pull request by sknot-rh**
Signed-off-by: Stanislav Knot <sknot@redhat.com>

### Type of change
- Bugfix

### Description
`compareDottedVersions` does not cope with versions like `3.0-IV1` so it cannot be used for comparing `inter.broker.protocol.version`. 
If the version contains `-` character, only its prefix is used. So if we provide `2.8-IV1`, the version `2.8` will be used.

Fixes https://github.com/strimzi/strimzi-kafka-operator/issues/6048

### Checklist
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [x] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaVersion.java [hunk 1]
```diff
@@ -423,6 +423,9 @@ public int compareVersion(String version) {
      * 1 if version1 &gt; version2.
      */
     public static int compareDottedVersions(String version1, String version2) {
+        // Version like X.Y-IVZ can get through validation
+        version1 = version1.split("-")[0];
+        version2 = version2.split("-")[0];
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaVersion.java:428]: Isn't the same method also used to compare the protocol versions?

**Comment by scholzj** [comment]
Maybe you can elaborate a bit more on:
* What is the problem and what happens today when it fails
* How does this fix it

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaVersionTest.java [hunk 2]
```diff
@@ -31,6 +31,20 @@ private Reader getKafkaVersionsReader(String kafkaVersions) {
         return new InputStreamReader(KafkaVersion.class.getResourceAsStream("/" + kafkaVersions), StandardCharsets.UTF_8);
     }
 
+    @ParallelTest
+    public void parsingInvalidVersionTest() {
```
Comment (scholzj) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaVersionTest.java:35]: I'm not sure I get the test => you added new method `compareDottedIVVersion`. But you do not seem to use it in the test. Plus the `compareDottedVersion` which I would expect to fail seems to be passing here for `2.8-IV1`.

**Review by scholzj — COMMENTED** [review]
The code looks good. But I do not really understand the test you added. I think it would be also good to add test using the IV versions in LMFV and IBPV into `KafkaUpgradeDowngradeMockTest`.

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaVersionTest.java [hunk 3]
```diff
@@ -31,6 +31,24 @@ private Reader getKafkaVersionsReader(String kafkaVersions) {
...
+
+        assertThrows(NumberFormatException.class, () -> {
+            KafkaVersion kvFail = new KafkaVersion("why", "you", "little", "3.6.9", false, true, "");
+            KafkaVersion.compareDottedIVVersions("2.8.0", kvFail.protocolVersion());
+        });
+
+        assertThrows(NumberFormatException.class, () -> {
+            KafkaVersion kvFail = new KafkaVersion("2.8.1-IV1", "you", "little", "3.6.9", false, true, "");
+            KafkaVersion.compareDottedVersions("2.8.0", kvFail.protocolVersion());
+        });
```
Comment (scholzj) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/model/KafkaVersionTest.java:48]: BTW: The `KafkaVersion` instances are managed by us. So they would not contain anything stupid. They will IMHO always be something like `new KafkaVersion("2.8.0", "2.8", "2.8", "3.6.9", false, true, "")`. It is the version you compare it to which will be usually user provided? So maybe the tests make more sense the other way around? Having a valid `KafkaVersion` and comparing it with invalid values?

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaVersion.java [hunk 4]
```diff
@@ -440,6 +440,20 @@ public static int compareDottedVersions(String version1, String version2) {
...
+     * Compare two decimal version strings, e.g. 3.0 &gt; 2.7-IV1. Ignores the IV part
+     * @param version1 The first version.
+     * @param version2 The second version.
+     * @return Zero if version1 == version2;
+     * -1 if version1 &lt; version2;
+     * 1 if version1 &gt; version2.
+     */
+    public static int compareDottedIVVersions(String version1, String version2) {
+        String trimmedVersion1 = version1.split("-")[0];
+        String trimmedVersion2 = version2.split("-")[0];
```
[the remark this answers is not in the document]
Comment (sknot-rh) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaVersion.java:453]: This should be already validated by `KafkaConfiguration.validate(KafkaVersion)`

[the remark this answers is not in the document]
Comment (tombentley) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/model/KafkaVersion.java:453]: Could you add a comment then?

**Review by tombentley — APPROVED** [review]

File: cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaUpgradeDowngradeMockTest.java [hunk 5]
```diff
@@ -211,6 +211,40 @@ public void testUpgradeWithMessageAndProtocolVersions(VertxTestContext context)
...
+    @Test
+    public void testUpgradeWithIVMessageAndProtocolVersions(VertxTestContext context)  {
+        String iv = "-IV0";
+        Kafka initialKafka = kafkaWithVersions(KafkaVersionTestUtils.PREVIOUS_KAFKA_VERSION,
+                KafkaVersionTestUtils.PREVIOUS_FORMAT_VERSION + iv,
+                KafkaVersionTestUtils.PREVIOUS_PROTOCOL_VERSION + iv);
+
+        Kafka updatedKafka = kafkaWithVersions(KafkaVersionTestUtils.LATEST_KAFKA_VERSION,
+                KafkaVersionTestUtils.PREVIOUS_FORMAT_VERSION,
+                KafkaVersionTestUtils.PREVIOUS_PROTOCOL_VERSION);
```
Comment (scholzj) [inline cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KafkaUpgradeDowngradeMockTest.java:225]: Should it use the `-IV0` version here as well? Essentially, but not using it here, you limit what is tested here. The test as it is right now suggests that the user has Kafka configured with for example:
* version 2.8.0
* IBPV 2.8-IV0
* LMFV 2.8-IV0

An suddenly changes to for example:
* version 3.0.0
* IBPV 2.8
* LMFV 2.8

This seems not likely.

**Review by scholzj — APPROVED** [review]
Thanks- LGTM assuming the regression tests pass.

**Pull request merged**